### PR TITLE
Multimodal dataset builder + docs

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

There is currently no way to load a custom multimodal dataset. The current workflow is users will write their own transform and builder - which is already a big lift - but on top of that our custom component imports don't even work (#1540). So basically, it's impossible to use anything other than llava instruct and cauldron. This sucks.

I added a quick `multimodal_chat_dataset` builder. It assumes conversational data with images, much like llava instruct and ShareGPT4V. To keep changes minimal, I only support ShareGPT format. We can expand to other chat formats or to instruct data later.

This unfortunately required me to generalize the LlavaInstructToMessages (cc @joecummings, I agree we shouldn't generalize too early but in this case I think it's warranted). Again, to keep changes minimal, I simply merged LlavaInstructToMessages with ShareGPTToMessages. There were identical in handling text, the only difference was the image loading logic.

With this builder, now I finally have something to point to and discuss in the multimodal dataset docs, and users have some avenue to finetune multimodal on their own data. A bit of extra work, but worthwhile imo.

#### Changelog
What are the changes made in this PR?
* Merge ShareGPTToMessages and LlavaInstructToMessages
* Merge any relevant tests
* Add `multimodal_chat_dataset` builder
* Add doc page on multimodal datasets

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/basics/multimodal_datasets.rst [hunk 1]
```diff
@@ -0,0 +1,285 @@
+.. _multimodal_dataset_usage_label:
+
+===================
+Multimodal Datasets
+===================
+
+Multimodal datasets include both text and image data used to train VLMs. torchtune currently supports multimodal chat datasets.
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:7]:
```suggestion
Multimodal datasets include more than one data modality, e.g. text + image + speech, and can be used to train transformers. Torchtune currently only supports multimodal chat datasets for Vision-Language Models (VLMs).
```

File: docs/source/basics/multimodal_datasets.rst [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+The primary entry point for fine-tuning with multimodal datasets in torchtune is the :func:`~torchtune.datasets.multimodal.multimodal_chat_dataset`
+builder. This lets you specify a local or Hugging Face dataset that follows the multimodal chat data format
+directly from the config and train your VLM on it.
+
+.. _example_multimodal:
+
+Example multimodal dataset
+--------------------------
+
+Here is an example of a multimodal chat dataset for a visual question-answering task.
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:18]: I would add something like: "Notice how <image> is a place holder for the file provided in image_path, so we can know where to insert the image tokens".

I also think that the example should contain more than one image, so people know how to gerenalize it.

File: docs/source/basics/multimodal_datasets.rst [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    # torch.Size([4, 3, 224, 224])
+
+.. code-block:: yaml
+
+    # In config
+    model_transform:
+      _component_: torchtune.models.flamingo
+      path: /tmp/Meta-Llama-3-8B-Instruct/original/tokenizer.model
+      prompt_template: torchtune.data.QuestionAnswerTemplate
+      max_seq_len: 8192
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:77]: oh, this is confusing, because in every other config, this is done in the tokenizer. Maybe we should explicitly say that model_transform takes the place of the tokenizer. A better solution might be to rename every config.tokenizer as model_transform, but i am not sure if i even like it.

Comment (felipemello1) [reply docs/source/basics/multimodal_datasets.rst:77]: dumb question: Why wouldnt "FlamingoTransform" just be "FlamingoTokenizer"? we are tokenizing now images and text

File: docs/source/basics/multimodal_datasets.rst [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+    dataset:
+      _component_: torchtune.datasets.multimodal.multimodal_chat_dataset
+      source: json
+      data_files: data/my_data.json
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:82]: can we add a dummy 2-example-file of what my_data.json should look like

File: docs/source/basics/multimodal_datasets.rst [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+      split: train
+      column_map:
+        dialogue: conversations
+        image_path: image
+      image_dir: /home/user/dataset/
+      image_tag: "<image>"
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:88]: noob question: will this ever break if the user adds <image> to their prompt, or <image> is a special token and the tokenizer understands if it comes from the user or the ChatML?

File: docs/source/basics/multimodal_datasets.rst [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+and the user-assistant conversations are in another column.
+
+.. code-block:: text
+
+    |  conversations                     | image        |
+    |------------------------------------|--------------|
+    | [{"from": "human", "value": "Q1"}, | images/1.jpg |
+    |  {"from": "gpt", "value": "A1"}]   |              |
+
+As an example, you can see the schema of the `ShareGPT4V dataset <https://huggingface.co/datasets/Lin-Chen/ShareGPT4V>`_.
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:104]: it seems that it forces single image per row, is that right? If so, can we make it explicit that this is a constraint of this format? Otherwise, can we add an example with more than 1 image?

File: docs/source/basics/multimodal_datasets.rst [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+
+Loading multimodal datasets from Hugging Face
+---------------------------------------------
+
+You simply need to pass in the dataset repo name to ``source``, which is then passed into Hugging Face's ``load_dataset``.
+For most datasets, you will also need to specify the ``split`` and/or the subset via ``name``.
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:111]: nice! super useful!

File: docs/source/basics/multimodal_datasets.rst [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+                    {"type": "image", "content": pil_image},
+                    {"type": "text", "content": msg["value"]},
+                ]
+            )
+    print(user_message.contains_media)
+    # True
+    print(user_message.get_media())
+    # [<PIL.Image.Image>]
+    print(user_message.text_content)
+    # What time is it on the clock?
```
Comment (felipemello1) [inline docs/source/basics/multimodal_datasets.rst:213]: very nice!

File: torchtune/data/_messages.py [hunk 9]
```diff
@@ -335,14 +336,22 @@ class ShareGPTToMessages(Transform):
         ]
 
     Args:
-        train_on_input (bool): whether the prompt should remain unmasked. Default: False
+        train_on_input (bool): whether the prompt should remain unmasked. For multimodal datasets, ``train_on_input``
```
Comment (felipemello1) [inline torchtune/data/_messages.py:339]: can we add why this is always false?

File: torchtune/data/_messages.py [hunk 10]
```diff
@@ -335,14 +336,22 @@ class ShareGPTToMessages(Transform):
...
+            is always False and this value is ignored. Default: False
         column_map (Optional[Dict[str, str]]): a mapping from the expected columns ("conversations")
             to the new column names in the dataset. Key should be "conversations" and value should
             be the new column name. If None, keep the default "conversations".
             Default is None.
         new_system_prompt (Optional[str]): if specified, prepend a system message. This can
             serve as instructions to guide the model response. Setting this will OVERRIDE any system
             messages already present in the dataset. Default is None.
+        image_dir (Optional[Path]): path to the directory containing the images that is prepended to all image
```
Comment (felipemello1) [inline torchtune/data/_messages.py:348]: nit: fine to ignore, maybe add an example?

File: torchtune/data/_messages.py [hunk 11]
```diff
@@ -335,14 +336,22 @@ class ShareGPTToMessages(Transform):
...
+            paths in the dataset. If None, assume images are available in current working directory or are located
+            on a remote url. For text-only,leave as None. Default is None.
+        image_tag (Optional[str]): placeholder tags in the text content of each message to be replaced by dictionaries
```
Comment (felipemello1) [inline torchtune/data/_messages.py:351]: nit: maybe add an example? Based on this alone, i have no clue of what you are talking about

File: torchtune/data/_messages.py [hunk 12]
```diff
@@ -384,12 +397,41 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
...
+                            {"type": "image", "content": pil_image},
+                            {"type": "text", "content": content},
+                        ]
+                    else:
+                        content = format_content_with_images(
+                            content,
+                            image_tag=self.image_tag,
+                            images=[pil_image],
+                        )
+                    image_loaded = True
```
Comment (felipemello1) [inline torchtune/data/_messages.py:430]: what if there is more than one image per sample? wouldnt this break?

File: torchtune/data/_messages.py [hunk 13]
```diff
@@ -384,12 +397,41 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
...
+        image_loaded = False
         for message in sample[self._column_map["conversations"]]:
             role = role_map[message["from"]]
+            content = message["value"]
             if role == "system" and self.new_system_prompt is not None:
                 continue
-            content = message["value"]
-            masked = (role != "assistant") and (not self.train_on_input)
+            if role == "user":
+                if is_multimodal and not image_loaded:
```
Comment (felipemello1) [inline torchtune/data/_messages.py:413]: I dont understand this. After we load one image, then if there is some other message with role=="user", we will just ignore it? Shouldnt we raise an error instead of silently ignoring it?

File: torchtune/data/_messages.py [hunk 14]
```diff
@@ -384,12 +397,41 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
...
+                        ]
+                    else:
+                        content = format_content_with_images(
+                            content,
+                            image_tag=self.image_tag,
+                            images=[pil_image],
+                        )
+                    image_loaded = True
+
+            masked = (role != "assistant") and (
```
Comment (felipemello1) [inline torchtune/data/_messages.py:432]: dont fully follow it, could you please add a comment?

File: torchtune/datasets/multimodal/_llava_instruct.py [hunk 15]
```diff
@@ -5,125 +5,19 @@
...
-                        images=[pil_image],
-                    )
-                    image_loaded = True
-            masked = role != "assistant"
-            messages.append(Message(role=role, content=content, masked=masked))
-
-        return {"messages": messages}
-
-
 # TODO: point to Flamingo model transform as an example
```
Comment (felipemello1) [inline torchtune/datasets/multimodal/_llava_instruct.py:15]: do we need to address this TODO?

File: torchtune/datasets/multimodal/_multimodal.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+    *,
+    source: str,
+    column_map: Optional[Dict[str, str]] = None,
+    new_system_prompt: Optional[str] = None,
+    image_tag: Optional[str] = None,
+    image_dir: Optional[str] = None,
+    **load_dataset_kwargs: Dict[str, Any],
+) -> SFTDataset:
+    """
+    Configure a text+image dataset with conversations between user and model assistant.
```
Comment (felipemello1) [inline torchtune/datasets/multimodal/_multimodal.py:26]: nit: It bothers me a bit that its called multimodal, but it actually means text+image. Not asking for changes, just sharing

File: torchtune/datasets/multimodal/_multimodal.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+    as an alternative to :class:`~torchtune.datasets.SFTDataset`, as it is made to be config friendly.
+
+    The dataset is expected to follow the ShareGPT format:
+
+    .. code-block:: text
+
+        |  conversations                     | image        |
+        |------------------------------------|--------------|
+        | [{"from": "human", "value": "Q1"}, | images/1.jpg |
+        |  {"from": "gpt", "value": "A1"}]   |              |
```
Comment (felipemello1) [inline torchtune/datasets/multimodal/_multimodal.py:38]: same as before, if we only support one image, can we make it explicit? otherwise, can we have an example with 2 images?

**Review by felipemello1 — APPROVED** [review]
very nice! thanks for putting this up. I didnt have time to finish reviewin "multimodal_chat_dataset". Approving to unblock.

File: torchtune/datasets/multimodal/_multimodal.py (hunk 16 above)
Comment (joecummings) [reply torchtune/datasets/multimodal/_multimodal.py:26]: I concur.

File: docs/source/basics/multimodal_datasets.rst (hunk 4 above)
Comment (RdoubleA) [reply docs/source/basics/multimodal_datasets.rst:82]: that's shown above?

File: docs/source/basics/multimodal_datasets.rst (hunk 5 above)
Comment (RdoubleA) [reply docs/source/basics/multimodal_datasets.rst:88]: yes it will leade to unexpected behaviors, the burden is on the user for setting this correctly.. I don't know if there's an easy way to check for those edge cases.

File: docs/source/basics/multimodal_datasets.rst (hunk 6 above)
Comment (RdoubleA) [reply docs/source/basics/multimodal_datasets.rst:104]: yes, will state that it's constrained to single image per row for now

File: torchtune/data/_messages.py (hunk 9 above)
Comment (RdoubleA) [reply torchtune/data/_messages.py:339]: I'm not sure really, I think it's because it's not functionally useful? @pbontrager

File: torchtune/data/_messages.py (hunk 13 above)
Comment (RdoubleA) [reply torchtune/data/_messages.py:413]: this will only load and append image tokens for the first user message. all other messages, including other user messages, will be added as normal.

File: torchtune/datasets/multimodal/_llava_instruct.py (hunk 15 above)
Comment (RdoubleA) [reply torchtune/datasets/multimodal/_llava_instruct.py:15]: after certain events have transpired

File: torchtune/data/_messages.py (hunk 13 above)
Comment (felipemello1) [reply torchtune/data/_messages.py:413]: Do you mind making this comment in the code?
